remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,768 @@
+  slot_aamds = aamds;
+}
+
+void AnalyzeAliases(llvm::Module *module, const std::vector<StateSlot> &slots) {
+  const AAMDInfo aamd_info(slots, module->getContext());
+  llvm::DataLayout dl = module->getDataLayout();
+  auto bb_func = BasicBlockFunction(module);
+  for (auto &func : *module) {
+    if (&func != bb_func && !func.isDeclaration() &&
+        func.getFunctionType() == bb_func->getFunctionType()) {
+      auto sp = LoadStatePointer(&func);
+      ForwardAliasVisitor fav(slots, &dl, sp);
+      LOG(INFO) << "Analyzing aliases of func " << func.getName().str();
+      for (auto &block : func) {
+        for (auto &inst : block) {
+          fav.AddInstruction(&inst);
+        }
+      }
+      // if the analysis succeeds for this function, add the AAMDNodes
+      if (fav.Analyze()) {
+        LOG(INFO) << "Offsets: " << fav.offset_map.size();
+        LOG(INFO) << "Aliases: " << fav.alias_map.size();
+        LOG(INFO) << "Excluded: " << fav.exclude.size();
+        AddAAMDNodes(fav.alias_map, aamd_info.slot_aamds);
+        // Perform live set analysis
+        LiveSetBlockVisitor LSBV(func, slots, aamd_info.slot_scopes, fav.offset_map, bb_func->getFunctionType(), &dl);
+        LSBV.Visit();
+        // repeat, but now ready to remove
+        LSBV.RemoveDeadStores();
+      }
+    }
+  }
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(llvm::Function &func_,
+    const std::vector<StateSlot> &state_slots_,
+    const ScopeMap &scope_to_offset_,
+    const ValueToOffset &val_to_offset_,
+    const llvm::FunctionType *lifted_func_ty_,
+    const llvm::DataLayout *dl_)
+  : func(func_),
+    val_to_offset(val_to_offset_),
+    scope_to_offset(scope_to_offset_),
+    state_slots(state_slots_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove(),
+    lifted_func_ty(lifted_func_ty_),
+    func_used(),
+    on_remove_pass(false),
+    dl(dl_) {
+      for (auto &block_it : func) {
+        auto block = &block_it;
+        auto succ_block_it = successors(block);
+        if (succ_block_it.begin() == succ_block_it.end()) {
+          curr_wl.push_back(block);
+        }
+      }
+    }
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    for (auto block : curr_wl) {
+      if (VisitBlock(block)) {
+        // Updates have been made
+        // Add predecessors to next rotation
+        for (llvm::BasicBlock *pred : predecessors(block)) {
+          next_wl.push_back(pred);
+        }
+      }
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
```

Comment:
Make `next_wl` a local variable of this function, rather than a field of the class. It's not needed anywhere else.

Reply:
This is my favourite kind of refactoring.